### ISSUE TITLE
[debops.etherpad] upgrading (or downgrading) issues

### ISSUE THREAD

**Issue by myrdd** [debops/debops#422]
I've just performed an upgrade of an _etherpad-lite_ machine managed by `debops.etherpad`, upgrading from `1.6.6` to `1.7.0`. However, changing the `etherpad_version` variable was not sufficient. I had to…

* remove the `'{{ etherpad_src_dir + "/" + etherpad_repository }}'` directory to fulfill the `etherpad_register_source is changed` condition (used by 3 subsequent tasks),
* remove the `'{{ etherpad_home }}/.node-gyp'` directory created by Etherpad's `bin/installDeps.sh` script (which should be executed again when `etherpad_register_checkout is changed`).

In other words:

```
rm -rf /usr/local/src/etherpad-lite/etherpad-lite/ /var/local/etherpad-lite/.node-gyp/
```

The bug I mainly want to report is the latter removal. The task creating the `.node-gyp` directory looks like this:

https://github.com/debops/debops/blob/dcfce3d296bb994743b17a865a8315cf1bf51ffa/ansible/roles/debops.etherpad/tasks/main.yml#L123-L130

I think the `creates:` line should be removed because, AFAIU, as soon as `.node-gyp` exists, the task won't be run again.

Is there some solution for the first „bug“?

**Comment by drybjed** [comment]
Thanks for the bug report. The code is downloaded via `git` and the `etherpad_register_source` variable should contain variables that provide the commit id before and after the checkout - this could be used as a trigger to perform any updates. See for example how [`debops.gitlab` uses them to trigger the checkout after the source is cloned](https://github.com/debops/debops/blob/master/ansible/roles/debops.gitlab/tasks/download_gitlab_ce.yml#L61-L73).

**Comment by myrdd** [comment]
(I'm not sure if this comment should be covered in a separate issue. If so, just tell me.)

I know that using conditions based on `etherpad_register_source` would follow the style of other debops roles, such as the gitlab one. However, may I suggest a different approach using `facts.d`?

```yml
- name: create facts.d
  file:
    path: '/etc/ansible/facts.d'
    state: 'directory'

# ...

- name: Clone Etherpad source code
- name: Check if Etherpad is checked out  # can IMHO be removed with this proposed change
- name: Create Etherpad directory
- name: Prepare Etherpad worktree

- name: Get commit hash of target checkout
  command: git rev-parse {{ etherpad__version }}
  # ...
  register: etherpad_register_version_commithash

- name: Checkout Etherpad
  command: git checkout --force {{ etherpad_version }}
  # ...
  register: etherpad_register_checkout
  when: (ansible_local.debops_etherpad_checkout is undefined or
         ansible_local.debops_etherpad_checkout.commit is undefined or
         ansible_local.debops_etherpad_checkout.commit !=
             etherpad_register_version_commithash.stdout)

- name: update facts.d (checkout commit)
  copy:
    dest: '/etc/ansible/facts.d/debops_etherpad_checkout.fact'
    content: '{ "commit": "{{ etherpad_register_version_commithash.stdout }}" }'

- name: Generate Etherpad session key
- name: Generate Etherpad configuration
- name: Create log directory

- name: Install Etherpad dependencies
  command: bin/installDeps.sh
  # ...
  when: (etherpad_register_checkout is changed or
         ansible_local.debops_etherpad_installation is undefined or
         ansible_local.debops_etherpad_installation.commit is undefined or
         ansible_local.debops_etherpad_installation.commit !=
             etherpad_register_version_commithash.stdout)

- name: update facts.d (installation commit)
  copy:
    dest: '/etc/ansible/facts.d/debops_etherpad_installation.fact'
    content: '{ "commit": "{{ etherpad_register_version_commithash.stdout }}" }'

# ...
```

* the `debops_etherpad_checkout.commit` local fact will hold the **commit hash** corresponding to the checkout
* the `debops_etherpad_installation.commit` local fact will hold the **commit hash** corresponding to the **installation**
* both the "checkout" and the "install" tasks will be run _only if necessary_

**The main reason** why I favor the usage of `facts.d` is to be able to catch installation failures. Especially with Node projects I often get errors on installation. Using my above approach, in case the installation fails, the "install" task will definitely be run on the next playbook run. See these two cases:

* In the simple case, the "install" task will be triggered by `ansible_local.debops_etherpad_installation.commit != etherpad_register_version_commithash.stdout`.
* `etherpad_register_checkout is changed` will trigger in the following rare but possible case — steps to reproduce:
  * change etherpad_version
  * run playbook
  * upgrade fails
  * change etherpad_version version back to its previous value
  * run playbook again

**Comment by drybjed** [comment]
Going with Ansible facts is probably a good idea to trigger an upgrade/downgrade, but I wouldn't want to statically record old and current `git` commit in them, especially in the way you propose. The `debops.etherpad` role already has an Ansible fact script, perhaps you could extend it to get the current `git` branch/tag of the checked out Etherpad installation, then in the role compare the existing one with the version the user set in the inventory - if they differ, perform the upgrade. Also, the script could check if Etherpad is installed correctly - maybe there's a way to detect correctly installed dependencies? Then you can set the `ansible_local.etherpad.installed` variable to `True`/`False` and this can then be used by the role to decide if it should try and reinstall Etherpad again if something failed before.

**Comment by myrdd** [comment]
I have overseen the `etc/ansible/facts.d/etherpad.fact.j2` file, of course such a script is a better solution than the way I proposed.

> perhaps you could extend it to get the current `git` branch/tag of the checked out Etherpad installation

Indeed. With `PWD=/usr/local/src/etherpad-lite/etherpad-lite/` I can run `git rev-parse HEAD`, it works.

> then in the role compare the existing one with the version the user set in the inventory - if they differ, perform the upgrade

yes

> maybe there's a way to detect correctly installed dependencies?

FTR: `bin/installDeps.sh`: https://github.com/ether/etherpad-lite/blob/develop/bin/installDeps.sh
